cmds/id/id.go

```diff
@@ -0,0 +1,265 @@
+	if err != nil {
+		return err
+	}
+
+	var passwdInfo []string
+
+	passwdScanner := bufio.NewScanner(passwdFile)
+
+	for passwdScanner.Scan() {
+		passwdInfo = strings.Split(passwdScanner.Text(), ":")
+		if val, err := strconv.Atoi(passwdInfo[2]); err != nil {
+			return err
+		} else if val == u.GetUid() {
+			u.Name = passwdInfo[0]
+			return nil
+		}
+	}
+
+	return fmt.Errorf("User is not in %s", PASSWD_FILE)
+}
+
+func (u *User) GetGroups() map[int]string {
+	return u.Groups
+}
+
+func (u *User) GetGidName() string {
+	val := u.GetGroups()[u.GetUid()]
+	return val
+}
+
+// pullGroups aggregates the groups that the User is in.
+func (u *User) pullGroups() error {
+	groupsNumbers, err := syscall.Getgroups()
+	if err != nil {
+		return err
+	}
+
+	groupsMap, err := readGroups()
+	if err != nil {
+		return err
+	}
+
+	for _, groupNum := range groupsNumbers {
+		if groupName, ok := groupsMap[groupNum]; ok {
+			u.Groups[groupNum] = groupName
+		} else {
+			return fmt.Errorf("Inconsistent %s file", GROUP_FILE)
+		}
+	}
+	return nil
+}
+
+// readGroups reads the GROUP_FILE for groups.
+// It assumes the format "name:passwd:number:groupList".
+func readGroups() (map[int]string, error) {
+	groupFile, err := os.Open(GROUP_FILE)
+	if err != nil {
+		return nil, err
+	}
+
+	var groupInfo []string
+
+	groupsMap := make(map[int]string)
+	groupScanner := bufio.NewScanner(groupFile)
+
+	for groupScanner.Scan() {
+		groupInfo = strings.Split(groupScanner.Text(), ":")
+		groupNum, err := strconv.Atoi(groupInfo[2])
+		if err != nil {
+			return nil, err
+		}
+
+		groupsMap[groupNum] = groupInfo[0]
+	}
+
+	return groupsMap, nil
+}
+
+// NewUser is a factory method for the User type.
+func NewUser() (*User, error) {
+	emptyMap := make(map[int]string)
+	u := &User{"", -1, -1, -1, emptyMap}
+	u.pullUid()
+	u.pullEuid()
+	u.pullGid()
+	if err := u.pullGroups(); err != nil {
+		return nil, err
+	}
+	if err := u.pullName(); err != nil {
+		return nil, err
+	}
+	return u, nil
+}
+
+// IDCommand runs the "id" with the current user's information.
+func IDCommand(u User) {
+	if !flags.G {
+		if flags.u {
+			if flags.n {
+				fmt.Println(u.GetName())
+				return
+			}
+			fmt.Println(u.GetUid())
+			return
+		} else if flags.g {
+			if flags.n {
+				fmt.Println(u.GetGidName())
+				return
+			}
+			fmt.Println(u.GetGid())
+			return
+		}
+
+		fmt.Printf("uid=%d(%s) ", u.GetUid(), u.GetName())
+		fmt.Printf("gid=%d(%s) ", u.GetGid(), u.GetGidName())
+	}
+
+	if !flags.G {
+		fmt.Print("groups=")
+	}
+	n := 0
+	length := len(u.Groups)
+	for gid, name := range u.Groups {
+
+		if !flags.G {
+			fmt.Printf("%d(%s)", gid, name)
+
+			if n < length-1 {
+				fmt.Print(",")
+			}
+			n += 1
+		} else {
+			if flags.n {
+				fmt.Printf("%s ", name)
+			} else {
+				fmt.Printf("%d ", gid)
+			}
+		}
+	}
+	fmt.Println()
+}
+
+func main() {
+	if err := initFlags(); err != nil {
+		l.Fatalf("id: %s", err)
+	}
+
+	if theChosenOne, err := NewUser(); err != nil {
```

Comment:
Normally, you'd structure this differently: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow